src/mesh/analysis.rs

```diff
@@ -193,14 +193,57 @@ pub fn border_vertex_indices(edge_sharing: &EdgeSharingMap) -> HashSet<u32> {
     border_vertices
 }
 
+#[derive(Debug, PartialEq)]
+pub enum BorderEdgeLoopsResult {
+    Found(Vec<Vec<OrientedEdge>>),
+    FoundWithNondeterminism(Vec<Vec<OrientedEdge>>),
+    Watertight,
+}
+
 /// Finds continuous loops of border edges, starting from a random edge.
 ///
 /// The mesh may contain holes or islands, therefore it may have an unknown
-/// number of edge loops. If two or more edge loops meet at a single vertex, it
-/// is undefined whether they end up in the result as separate border edge
-/// loops, or are joined into a single border edge loop.
-pub fn border_edge_loops(edge_sharing: &EdgeSharingMap) -> Vec<Vec<OrientedEdge>> {
+/// number of edge loops. If two or more edge loops meet at a single vertex, the
+/// result is non-deterministic: a random path of more possibilities is chosen
+/// to form an edge loop.
+pub fn border_edge_loops(edge_sharing: &EdgeSharingMap) -> BorderEdgeLoopsResult {
     let mut border_edges: Vec<_> = border_edges(edge_sharing).collect();
+
+    // If there are no border edges, the mesh is watertight.
+    if border_edges.is_empty() {
+        return BorderEdgeLoopsResult::Watertight;
+    }
+
+    // Check if the border edges are branching (the loops are touching at a
+    // single vertex). If they are, the result is non-deterministic.
+
+    // FIXME: @Optimization: use a Map structure or at least trim the list of
+    // vertices also from below. Find the maximum index that occurs in the
+    // border edge loop.
```

Comment:
Separate FIXME from explanatory comment.

Reply:
Also, not sure by what you mean by the FIXME. This does a single iteration, which is not too bad, considering what the rest of the algorithm does.

Reply:
the mesh may have millions of vertices, out of which last three are part of border edges. this then allocates a vector of million integers. I want to prevent that. the easiest thing is to find the first and the last border vertex index and only allocate such chunk of the list, but that is a very dull optimization